File: src/commands/query.ts
import { type Renderable, encode, summarizeComponent } from "../output/index.js";
import type { QueryResult } from "../output/index.js";
import { searchComponents, searchDocs } from "../search.js";
import {
  DEFAULT_QUERY_LIMIT,
  defineGlobalCommand,
  loadBundle,
  parseLimit,
  resolveFormat,
} from "./shared.js";

export const queryCommand = defineGlobalCommand({
  meta: { name: "query", description: "Search components and docs for a term" },
  args: {
    term: { type: "positional", description: "search term", required: true },
    limit: {
      type: "string",
      description: "maximum results per category (0 = all)",
      default: String(DEFAULT_QUERY_LIMIT),
    },
  },
  async run({ args }) {
    const format = resolveFormat(args);
    const bundle = await loadBundle(args);

    const term = args.term;
    const limit = parseLimit(args.limit);

    const result: QueryResult = {
      term,
      components: searchComponents(bundle.components.components ?? {}, term, limit).map((m) =>
        summarizeComponent(m.component, m.score),
      ),
      docs: searchDocs(bundle.docs.docs ?? {}, term, limit).map((m) => ({
        id: m.doc.id ?? "",
        title: m.doc.title ?? "",
        score: m.score,
      })),
    };
    if (bundle.warnings.length > 0) result.warnings = bundle.warnings;

    const payload: Renderable = { kind: "query", value: result };
    process.stdout.write(encode(format, payload));
  },
});

File: internal/cli/query.go
package cli

import (
	"os"

	"github.com/spf13/cobra"

	"github.com/adriankarlen/storyquery/internal/output"
	"github.com/adriankarlen/storyquery/internal/search"
)

const defaultQueryLimit = 10

func newQueryCmd(gf *globalFlags) *cobra.Command {
	var limit int
	cmd := &cobra.Command{
		Use:   "query <term>",
		Short: "Search components and docs for a term",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			format, err := resolveFormat(gf)
			if err != nil {
				return err
			}
			bundle, err := loadBundle(cmd.Context(), gf)
			if err != nil {
				return err
			}

			term := args[0]
			comps := search.Components(bundle.Components.Components, term, limit)
			docs := search.Docs(bundle.Docs.Docs, term, limit)

			res := output.QueryResult{Term: term, Warnings: bundle.Warnings}
			for _, m := range comps {
				res.Components = append(res.Components, output.SummarizeComponent(m.Component, m.Score))
			}
			for _, m := range docs {
				res.Docs = append(res.Docs, output.DocSummary{ID: m.Doc.ID, Title: m.Doc.Title, Score: m.Score})
			}
			return output.Encode(os.Stdout, format, res)
		},
	}
	cmd.Flags().IntVar(&limit, "limit", defaultQueryLimit, "maximum results per category (0 = all)")
	return cmd
}
